src/sql/src/pure/mysql.rs

```diff
@@ -215,3 +231,33 @@ pub(super) fn normalize_column_refs<'a>(
     seq.dedup();
     Ok(seq)
 }
+
+pub(super) async fn validate_requested_subsources_privileges(
+    requested_subsources: &[RequestedSubsource<'_, MySqlTableDesc>],
+    conn: &mut mz_mysql_util::MySqlConn,
+) -> Result<(), PlanError> {
+    // Ensure that we have correct privileges on all tables; we have to do this before we
+    // start snapshotting because if we discover we cannot `SELECT` from a table while
+    // snapshotting, we break the entire source.
+    let tables_to_check_permissions = requested_subsources
+        .iter()
+        .map(|RequestedSubsource { upstream_name, .. }| upstream_name_to_table(&upstream_name))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    validate_source_privileges(&mut *conn, &tables_to_check_permissions)
+        .await
+        .map_err(|err| match err {
+            MySqlError::MissingPrivileges(missing_privileges) => {
+                MySqlSourcePurificationError::UserLacksPrivileges(
+                    missing_privileges
+                        .into_iter()
+                        .map(|mp| (mp.privilege, mp.qualified_table_name))
+                        .collect(),
+                )
+                .into()
+            }
+            _ => PlanError::MySqlConnectionErr { cause: err.into() },
+        })?;
+
+    Ok(())
+}
```

Comment:
Is there a race condition here? If this succeeds and then our permissions are revoked before starting the snapshot, then will the source be broken?

Reply:
Yeah then the source will be created but go into an error state during the snapshot phase and result in a `stalled` status. There's not much we can do to prevent this and it's the same way the postgres source operates now (check permissions during purification, hope they dont change later).